app/http/handler_auth.go

```
@@ -0,0 +1,102 @@
+package http
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/fredcarle/int-fullstack-fred-api/app"
+	"github.com/pkg/errors"
+)
+
+type loginRequest struct {
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
+
+func (h *Handler) login(ctx *Context) {
+	if ctx.r.Method != "POST" {
+		h.Error(ctx.w, app.ErrNotFound.Wrap(errors.New("bad method")))
+		return
+	}
+
+	if ctx.session.GetToken() != "" {
+		u, err := ctx.session.VerifyAuthUser()
+		if err == nil && u != nil {
+			http.SetCookie(ctx.w, &http.Cookie{
+				Name:   "auth-token",
+				Value:  ctx.session.GetToken(),
+				MaxAge: int(time.Hour * 24),
+				Path:   "/api/v1/",
+			})
+			h.sendJSON(ctx.w, u, http.StatusOK)
+			return
+		}
+	}
+
+	lr := loginRequest{}
+	if err := getJSON(ctx.r, &lr); err != nil {
+		h.Error(ctx.w, err)
+		return
+	}
+
+	if lr.Email == "" || lr.Password == "" {
+		h.Error(ctx.w, app.ErrBadRequest)
+		return
+	}
```

Why check these here? Just try them and let them fail in `GetFromEmail` or `HashMatch`?

---

I think it's useful in the case where we would want the frontend to show a different error than a failed `GetFromEmail`. But to your point, that situation could also be handled by validations in the frontend before the request is ever sent.